mspnp/reference-architectures
# Deploy ingress controller

closes: #199296

out of scope: HPA

- assign ntw contrib role to the cluster system-assigned id because in the current Managed Identity model, only system assigned identities are supported.
- deploy Traefik version v1.7.20 
- node selection constraint. Select user nodepool for ingress controller. This is under the premise that everything we bring to the cluster should land in user nodepools
- design review feedback: HA 
   * Ingress Controller: 
      1. match the uninteresting workload number replicas 
      1. prevent from co-locating replicas of traefik on a single node
   * Uninteresting Workload:
      1. prevent from co-locating replicas of uninteresting workload on a single node
      1. inform the scheduler that all its replicas are desired to be co-located with traefik pods

```
  System        User          User
  Node          Node 1        Node 2

+---------+  +----------+  +----------+
|   ...   |  |traefik|1 |  |traefik|2 |
+---------+  +----------+  +----------+
             +----------+  +----------+
             |workload|1|  |workload|2|
             +----------+  +----------+
```
- move uninteresting workload to the root folder
- upgrade traefik from v1.7.20 to v2.2.1
  * use kubernetes ingress provider
  * expose health and http(s) endpoints
  * disable stats collection in production workloads
  * add securecontext
  * standarize labels
  * config traefik using static file with configmap
  * migrate config
- design review feedback: [TLS] configure Traekif and workload route with TLS
  * move traefik to the a0008 ns
  * create wildcard default self signed certificate: *.bicycle.contoso.com
  * create secret with tls cert
  * add default certificate mounting volumes from secrets
  * configure unintersting workload ingress resource to use tls without
    secretref using *.bicycle.contoso.com cert to serve the specific route
  * middlewares to compress using gzip + move permantely (Http308) to https schema
    http client requests
  * narrative assets:
    1. prescribe steps to check Traefik can handle HTTPS request even
      before creating an ingress resource (router)
    1. prescribe steps to check the ingress resource object state
    1. prescribe steps to check router, ssl offloading and http redirect
      to https with the uninteresting workload
    1. add missing pre-req
```
                                     |            AKS             |
bicycle.contoso.com                --|   Internal Load Balancer   |-> <returns 404 but has a matching certificate>
                                     |        10.10.4.X           |
bu0001a0008-00.bicycle.contoso.com --| Traefik Ingress Controller |-> aspnetweb-service:80
                                     |           Https            |
```
- design review feedback: [App Gw] cert integration this is about start integrating the Azure Application Gateway with the AKS Ingress Controller (Traefik)
  * output az app gw name from cluster stamp
  * add new Azure Application Gateway section to the README (future
  narrative home)
  * upload root cer to az app gw
  * configure default http settings
